# Python Basics--Beginner Notes

 Learn python step-by-step in a simple way!

# 1) What is python?

In [ ]:
Python is a powerful and easy programming language. we use it for:
 * Data analysis
 * Machine learning
 * Web development
 * Automation
Let's start coding!

In [8]:
 # This is your first python program
print("Hello,Aryan!")

Hello,Aryan!


# 2) Variable and Data types

Variables store information

In [4]:
# Variables store information
name = "Aryan"
age = 16
height = 6.1
# print variables
print("name:",name)
print("age:",age)
print("height:",height)

name: Aryan
age: 16
height: 6.1


# 3) Operators

We can do basic math using operators.

In [5]:
# Arithmetic operations
a = 10
b = 123

print("addition:", a + b)
print("sunbtraction:", a - b)
print("multiplication:", a * b)
print("division:", a / b)

addition: 133
sunbtraction: -113
multiplication: 1230
division: 0.08130081300813008


# 4) Comments

Comments help explain the code and are ignored by Python.

In [10]:
# Anthing after # is a comment
# Python ignore it when running code
print("Comments make code easy to understand!")

Comments make code easy to understand!


In [11]:
a=2
b=4
print(a+b)

6


# Variables introduction: String,Numbers,boolean

Variables are containers that store some type of data. for example soda can is a container that contains soda.

In [12]:
x="soda"
x

'soda'

In [13]:
print(x)

soda


In [14]:
y="beans"
y

'beans'

In [15]:
print(y)


beans


In [16]:
type(y)

str

In [17]:
z= 100
z

100

In [18]:
print(z)


100


In [19]:
type(z)


int

In [20]:
p=20.5
q=100

In [21]:
total=z+p+q

In [22]:
total

220.5

In [23]:
print(total)

220.5


In [24]:
type(total)

float

In [26]:
a=total>500
a

False

In [27]:
type(a)

bool

Integer is a whole number with no decimal point.eg 34,12,500
Float is a number with decimal point e.g. 1.34,0.45,100.5
string data type store text information eg. suresh
Boolean type is used to store true or false value
### Everything is an object in object in python--Simple Explanation
Imagine python is a big a toy shop
In that toy shop:
Every toy is an object
Each toy has:
a name
some data
things it can do(actions)
Python words the same way!
# In Python,Everything is an object
This means:
Numbers
String
Lists
Tuples
Funcions
Even your code itself
Even True/False
Even None
All of them are objects


In [28]:
foo = 'jalebi'
foo

'jalebi'

In [29]:
bar= foo
foo

'jalebi'

In [30]:
id(bar)

2119275767648

In [31]:
foo

'jalebi'

# Variable Naming Rules

# 1) Can't use reserved keyboards to store variable name


###In Python, some words are already used by the language for
special purposes. These are called reserved keywords,like:for,if,else,
while,class,return,true,false,none,try,except you cannot use these words as variable names.

In [1]:
# wrong:
for = 10
class = "hello"

SyntaxError: invalid syntax (1751472206.py, line 2)

In [2]:
# correct:
number = 10
satish = 3
greeting ="hello"
print(number)
print(satish)

10
3


### Python Reseved Keywords (Cannot Be Used as Variable Names)

In [12]:
# You cannot use any of these as variable names.
# ❌ Wrong:

for = 20

SyntaxError: invalid syntax (2210437058.py, line 4)

In [18]:
# ✔ Correct:
for_loop_count = 20
print("for_loop_count")
print(20)

for_loop_count
20


#🧠 Python keywords - Examples for each one

In [22]:
# False represents a boolean false value
x = "false"
print(x)

false


In [23]:
# None means "no value" or "empty"
x = "none"
print(x)

none


In [27]:
# True represents a boolean true values
x = "true"
print(x)

true


In [30]:
# and returns true only if both sides are true
print("true" and "false") #false

false


In [31]:
# as gives an an alias (short name) to modules
import math as m
print(m.sqrt(16))

4.0


In [32]:
# assert checks if condition is true, otherwise error
x = 5
assert x > 0 # no error

In [33]:
# asyync is used to define asynchronous functions
async def hello():
    return "Hi!"

# 🧠 Understanding async in Python(Super Simple Explanation)

In [ ]:
Think like this:
Imagine you tell your computer:
"Do this work,but don't wait...continue doing other work too."
This is called asynchronous work (Working on more than one thing at the same time)

# 🧸 What does async def means?

async def is used to make a special type of funtion called a coroutine.
A coroutine is a function that:
 * Doesn't run immediately
 * Runs later when you ask it
 * Can pause and resume
 * Help Python do multiples tasks together(without waiting)

# ✔ Your code:


In [ ]:
# async is used to define asynchronous functions
async def hello():
    return "hi!"

# What this means:

* async="this function is special"
* def hello():=normal function name
* rreturn "Hi!"=what the function will give back

In [ ]:
BUT...
 You cannot do:
 hello()
 Because it's a coroutine.
 You must run it using:
 await hello()
 Or inside an async program.


# 🍦Simple Example(Story)
Imagine you ask your mom:

"Mummy,please boil water!"

But boiling water taskes time.

So your mom DOES NOT wait next to the gas.

She starts:
  * washing dishes
  * cleaning
  * cokking

while water is boiling

This is async.

 # Simple async example:
import asyncio

async def hello():
    return "Hi!"

async def main():
    msg = await hello()
    print(msg)

asyncio.run(main())

Output:
Hi!

In [5]:
import asyncio
async def hello():
    return "Hi!"

async def main():
    msg = await hello()
    print(msg)

asyncio.run(main())

RuntimeError: asyncio.run() cannot be called from a running event loop

# You're getting this error:
RuntimeError: asyncio.run() cannot be called from a runnig event loop
This happens only in Jupyter Notebook/Google Colab,not in normal Python.
Let me explain why,in a very simple way

# 🧠 Why the Error Happens?

Jupyter Notebook is already running an async loop in the background.

So when you write:

asyncio.run(main())

Python says:


"Hey! I'm already running an event loop.You can't start another one."

So it gives RuntimeError.

# 🟢 ✔ Correct Way to Run async Code in Jupyter Notebook

## **Method 1: Just use `await` directly**

No need for `asyncio.run()` in Jupyter.

### Example:

```python
async def hello():
    return "Hi!"

msg = await hello()
print(msg)
```

### Output:

```
Hi!
```

Easy! ✅


# 🟢 Method 2:Use this style for bigger programs
```python
import asyncio
async def main():
    print(await hello())

await main()
```

# ❌wrong in Jupyter:
asyncio.run(main()) # causes the error

# ✔Why await works directly in jupyter?
Because Jupyter Notebook alredy supports async.So you can directly say:
await something
without starting a new event loop.

## 🧸  Summary(Super Simple)
* Normal Python = use asyncio.run()
* Jupyter Notebook = use await directly

In [1]:
import asyncio
async def hello():
    return "Hi!"
async def main():
    msg = await hello()
    print(msg)
await main() # use await instead of asyncio.run()

Hi!


In [2]:
# break stops the loop immediately
for i in range(5):
    if i == 3:
        break
    print(i)

0
1
2


In [ ]:
# class is used to create custom objects
class dog:
    pass

# 🧠 What does this code mean?
```python
# class is used to create custom objects
class dog:
    pass
```
let's break it down step by step

## 🧸1.What is a class?(very simple)
A class is like a blueprint or design.

Example:
* A blueprint for a house
* A design for a toy
* A recipe for cake
from the bluerprint,you can make many  houses.from the recipe you can make many cakes.
Similarly,

    A class allows you to create many  objects.

# 🐶2.What is this "Dog"class?
class Dog: means:

"I am creating a design for a Dog."
But right now, the desidn is empty.

# 📝3.What does pass mean?
pass means:

"Don't do anything now."
Python needs something  inside the class.if you don't want to add anything yet,you write pass.

It is like saying:

"Teacher,page khali rakho.Baad mein likhenge."

# 🏗 4.What will we use this class for?
Later,we can create objects from this class:

d1 = Dog() # Dog object

d2 = Dog() # another Dog object

Now we have two dog objects.

# ⭐ A better example:

Let's add some features:
    
class Dog:
     
     def bark(self):
     
      print("Woof!")
Now:


mydog = Dog()

mydog.bark()

output:

Woof!

# 🎯 Summary (Very Simple)

| Word       | Meaning                        |
| ---------- | ------------------------------ |
| **class**  | Blueprint/Design               |
| **Dog**    | Name of the blueprint          |
| **pass**   | "Do nothing for now"           |
| **object** | Actual dog created from design |




In [1]:
# Empty class using pass
class Cat:
    pass

# Creating an object
c = Cat()
print(c)

In [2]:
class car:
    # Funciton inside class 
    def start(self):
        print("car is starting")
# Create object
my_car = car()
my_car.start()

car is starting


In [6]:
class Fruit:
    pass

# Creating objects from class
mango = Fruit()
apple = Fruit()
banana = Fruit()

print(mango, apple, banana)

<__main__.Fruit object at 0x000001D69E455D30> <__main__.Fruit object at 0x000001D69FBFFED0> <__main__.Fruit object at 0x000001D69FBFF890>


#Great question! But the truth is: You cannot fix or control a memory address in Python.
Python does not allow you to choose or lock a memory address.

Let me explain in very simple way.

# 🧠 WHY YOU CANNOT FIX MEMORY ADDRESS

when you create an object:
```python
x = 10
```

Python decides **automatically** where to store it in memory.

Just like:
   * You cannot choose your house number in big apartment
   * The building manager assigns it
Similarly:
 Python assign a random,available memory house number to each object.

You cannot change it.

# Why Python does not allow fixed addresses?
because:
 * It would be unsafe
 * It can crash your program
 * It breaks python's memory management
 * Pyhton automatically cleans(garbage collection)
Python handles everything  to keep you safe.

# 🧩 BUT -- You can make two variables share the same memory address

# EXAMPLE

a=10

b=a

 # now:
id(a)

id(b)

Both will be the same because b point to the same object.

# 🧩 You can also keep same memory address for strings
x = "hello"

y = "hello"

Now:

id(x)

id(y)

same memory address -- because Python reuses small strings.

# 🧠 Summary

| Question                              | Answer   |
| ------------------------------------- | -------- |
| Can you fix a memory address?         | ❌ No     |
| Who decides the memory address?       | ✔ Python |
| Can two variables share same address? | ✔ Yes    |
| Can we change object’s address?       | ❌ No     |

---



 # 🟢  Method 1: Using str(simple)


In [8]:
class Fruit:
    def __init__(self, name):
        self.name = name

    def __str__(self):
        return f"This is a {self.name} object"

mango = Fruit("Mango")
apple = Fruit("Apple")
banana = Fruit("Banana")

print(mango)
print(apple)
print(banana)

This is a Mango object
This is a Apple object
This is a Banana object


#  Of course! I'll explain this code step-by-step in very simple language,like you're in 3rd standard.

# 🟢 Full code

In [9]:
class fruit:
    def __init__(self,name):
        self.name = name

    def __str__(self):
        return f"This is a {self.name} object"

mango = fruit("mango")
apple = fruit("apple")
banana = fruit("banana")

print(mango)
print(apple)
print(banana)

This is a mango object
This is a apple object
This is a banana object


## Step-by-step Explanation

# 1) class Fruit:
You are creating a blueprint named fruit.
Think of it like: A drawing that expalins how to make a fruit.

# 2) def init(self,name):
This is the constructor.
It runs automatically when you create an object.
Like: When you make a new fruit = it immediately gets a name.

Example:
mango = fruit("mango")

here "mango" goes to name.

# 3) self.name = name
Meaning: Store the name inside the object.
like:
   * Mango object gets name"Mango"
   * Apple object gets name"Apple"
   * Banana object gets name"Banana"
     
self = means this object itself.

# 4) def str(self):
This tells Python:"What should I print when someone prints the object?"
Normal objects print:

```
<__main__.Fruit object at 0x123ABC>
```
But this function change it to something nice.
```

# 5) return f"This is a {self.name}object"
This builds a text message.
for mango object:

This is mango object

for apple:

This is an apple object

for banana:

This is a Banana object

# 6) Creating Objects
mango = fruit("Mango")

apple = fruit("Apple")

banana = fruit("banana")

You now create three fruit objects.

Each one gets a different naem.

# 7) Printing objects

print(mango)

print(apple)

print(banana)

This automatically calls the __str__ function.

So you get:

 OUTPUT:

This is a mango object

This is an apple object

This is a banana object

---

# 🎉 Final Simple Summary

| Code Part      | Meaning                |
| -------------- | ---------------------- |
| class Fruit    | Blueprint              |
| **init**       | Give each fruit a name |
| self.name      | Store name             |
| **str**        | Nice printing          |
| Fruit("Mango") | Create mango fruit     |
| print(mango)   | Show nice message      |

---



# 🟢 Method 2: Using repr (also works)


In [3]:
class Fruit:
    def __init__(self, name):
        self.name = name

    def __repr__(self):
        return f"Fruit(name={self.name})"

mango = Fruit("Mango")
apple = Fruit("Apple")
banana = Fruit("Banana")

print(mango)
print(apple)
print(banana)

Fruit(name=Mango)
Fruit(name=Apple)
Fruit(name=Banana)


#Sure! I will explain this code step-by-step, in s very simple way so you understand everything.

# 🧠 Step-by-Step Explanation 


# 1) class fruit:
You are creating a blueprint called fruit.

This blueprint tells python:"Here is how a fruit object should look."
    

# 2) def init(self,name):
    This is the constructor.
    It runs automatically when you create a new fruit object.
    Example:
    mango = fruit("Mango")
     * "Mango" goes into name
     *  self.name = name saves it inside it indide the object
So;
  * mango.name ="Mango"
  * apple.name ="Apple"
  * banana.name = "Banana"


# 3) self.name = name
this line store the naem of the fruits inside each object.

self means this object.

# 4) def repr(self):
This function controls: How the object looks when printed OR shown in a  list.

__reper__is mainly for developers,not and users.


# 5) return f"fruit(name={self.name})"
This is the actual text python will show.

so when you print:

print(mango)

It prints:

fruit(name=Mango)

for apple:

fruit(name=Apple)

for banana::

fruit(name=Banana)

 This replaces the default ugly output like:
<__main__.fruit object at 0x00000123ABCDEF>
__repr__gives you a clean,readable result.

# 6) Creating Objects
mango = fruit("Mango")

apple = fruit("Apple")

banana = fruit("Banana")

Each line creates a new object with a name

# 7) print(mango)

This calls the __repr__ method.
so you get:

OUTPUT:
fruit(name=mango)

fruit(name=apple)

fruit(naem=banana)


# 🟩 🆚 `__str__` vs `__repr__`

| Method     | Purpose        | Output Style         |
| ---------- | -------------- | -------------------- |
| `__str__`  | For users      | Beautiful, friendly  |
| `__repr__` | For developers | Exact object details |


# 🎯 Final Simple Summary

| Line           | Meaning                  |
| -------------- | ------------------------ |
| `class Fruit`  | Create fruit blueprint   |
| `__init__`     | Give name to each fruit  |
| `self.name`    | Store name in object     |
| `__repr__`     | Decide how object prints |
| `print(mango)` | Show object as text      |

In [1]:
# continue skips the current loop iteration
for i in range(5):
    if i == 2:
        continue
    print(i)

0
1
3
4


In [2]:
# def cretes afunction
def add(a, b):
    return a + b

In [3]:
# del removes a variable from memory
x = 10
del x

In [4]:
# elif = else if
x = 5
if x == 3:
    print("three")
elif x == 5:
    print("five")

five


In [8]:
# else runs when if condition is false
if "false":
    print("Hi")
else:
    print("bye")

Hi


In [9]:
# except handles errors
try:
    1 / 0
except ZeroDivisionError:
    print("Cannot divide")

Cannot divide


In [10]:
# finally runs always (even if error happens)
try:
    x = 10
finally:
    print("This always runs")

This always runs


In [11]:
# for loop repeats code
for i in range(3):
    print(i)

0
1
2


In [12]:
# from imports specific items from a module
from math import pi
print(pi)

3.141592653589793


In [13]:
# global allows modifiying outer variable
x = 10
def change():
    global x
    x = 20
change()
print(x)

20


In [14]:
# if cehcks acondition 
if 10 > 5:
    print("true")

true


In [15]:
# import loads a module
import random
print(random.randint(1,5))


4


In [16]:
# in checks membership
print(3 in [1,2,3])

True


In [18]:
# is checks identity (same object in memory)
print("none" is "none")

True


<>:2: SyntaxWarning: "is" with 'str' literal. Did you mean "=="?
C:\Users\Aryan\AppData\Local\Temp\ipykernel_25304\1954087868.py:2: SyntaxWarning: "is" with 'str' literal. Did you mean "=="?
  print("none" is "none")


In [23]:
# lambda creates small anonymous functions
square = lambda x: x * x
print(square(4))

16


In [24]:
# nonlocal modifies variable from outer function
def outer():
    x = "outer"
    def inner():
        nonlocal x
        x = "inner"
    inner()
    print(x)
outer()

inner


In [28]:
# not reverses boolean value
print("not true")

not true


In [32]:
# or returns true if any one condition is true
print("true or false")

true or false


In [33]:
# pass means "do nothing"
def myfunc():
    pass

In [34]:
# raise manually triggers an error
# raise vlauerror("error message")

In [35]:
# return sends value back from a function
def f():
    return 5

In [36]:
# try tests code that my causes an error
try:
    x = 1/0

_IncompleteInputError: incomplete input (3160945134.py, line 3)

In [37]:
i = 0
while i < 3:
    print(i)
    i += 1

0
1
2


In [38]:
# with automatically handels open/close of resources
with open("test.txt","w") as f:
    f.write("hello")

In [39]:
# yield turns a function into agenerator
def gen():
    yield 1
    yield 2
    

# 2. Special characters are not allowed

# Here is a clean and simple markdown table showing allowed vs not allowed variable names in python.

# 🧠 Allowed vs Not Allowed variavle names in python

| Variable Name | Allowed? | Reason                    |
| ------------- | -------- | ------------------------- |
| `name`        | ✅ Yes    | Only letters              |
| `age1`        | ✅ Yes    | Letters + numbers         |
| `my_name`     | ✅ Yes    | Underscore allowed        |
| `_score`      | ✅ Yes    | Can start with underscore |
| `StudentAge`  | ✅ Yes    | Uppercase letters allowed |
| `a123`        | ✅ Yes    | Letters + digits          |


### ❌ **Not Allowed Examples**

| Variable Name | Allowed? | Reason                   |
| ------------- | -------- | ------------------------ |
| `my-name`     | ❌ No     | Hyphen (-) not allowed   |
| `na$me`       | ❌ No     | `$` special character    |
| `1name`       | ❌ No     | Cannot start with number |
| `age#`        | ❌ No     | `#` not allowed          |
| `my name`     | ❌ No     | Space not allowed        |
| `@price`      | ❌ No     | Cannot start with `@`    |
| `num%ber`     | ❌ No     | `%` not allowed          |

---

# 🧸 **Simple Reminder**

✔ Allowed:

* Letters (A–Z, a–z)
* Digits (0–9)
* Underscore (_)

❌ Not Allowed:

* Special symbols
* Spaces
* Starting with numbers

---

In [40]:
$name="dhaval"

SyntaxError: invalid syntax (1667250009.py, line 1)

In [42]:
pizza^ = 100

SyntaxError: invalid syntax (2224969571.py, line 1)

# 3 . Variable names must begin with letter or_

# Here is a clear explanation for the rule:
## 3. Variable names must begin with a letter or

 # 🧠 Rule:
In Python, variable must start with:
 a letter = (A-Z or a-z) OR an underscore
You cannot start a variable names with a number or special symbol.
         

# 🟢 Correct Examples(Allowed)

name = "Aryan"
student_age = 17
_age= 16
score = 98
myValue = 66
* starts with letters = OK
* starts with underscore = OK

# 🔴 **Wrong example (not allowed)**
```python
1name = "Aryan"    # starts with number ❌
@age = 16           # starts with special symbol ❌
$price = 18         # special symbol ❌
5value = 1800        # number ❌
```

# 🧸 **Simple Explanation (3rd class style)**

Think of variable names like **children's names**:

✔ You can name someone **Amit** or **_Amit**
❌ But you cannot name someone **1Amit** or **@Amit**

Python follows the same rule.


# 🧠 **Examples Table**

| Variable Name | Allowed? | Reason                   |
| ------------- | -------- | ------------------------ |
| `name`        | ✅ Yes    | Starts with letter       |
| `_name`       | ✅ Yes    | Starts with underscore   |
| `age1`        | ✅ Yes    | Starts with letter       |
| `1age`        | ❌ No     | Starts with number       |
| `@value`      | ❌ No     | Starts with symbol       |
| `$money`      | ❌ No     | Special char not allowed |




In [43]:
4you = "me"

SyntaxError: invalid decimal literal (1970908349.py, line 1)

In [44]:
$address="blah"

SyntaxError: invalid syntax (2488452656.py, line 1)

# 4. Remaining characters should be a to z OR A to Z
OR 0 to 9 OR_

# Here is a simple,clear explanation for:

### 4. Remaining characters should be a-z OR A-Z OR 0-9  OR_

# 🧠 Rule :
After the first character, the rest of the charcters in a vraiable naem can be:
small letters = a to z capital letters = A to z numbers = 0 to 9 underscore.
You cannot use any special characters.

# 🟢 Correct Example(Allowed)
name1
student_age
my_vlaue_2
_Age3
Hello_world 
python123

These work because they use only:
* letters
* digits
* underscores

#  🔴 **Wrong Examples (Not Allowed)**


my-name     # hyphen not allowed

age#1       # # not allowed

value@12    # @ not allowed

num%ber     # % not allowed

my name     # space not allowed

# 🧸 **Very Simple Explanation (3rd class style)**

Think of variable name like a **word**:

✔ You can use letters


✔ You can add numbers


✔ You can add underscores `_`

❌ You cannot add symbols like:

`@  #  $  %  !  -  *  /  space`

# 🧠 **Examples Table**

| Variable Name | Allowed? | Reason                    |
| ------------- | -------- | ------------------------- |
| `value_1`     | ✅ Yes    | Uses letters, digits, `_` |
| `num_ber_2`   | ✅ Yes    | All allowed characters    |
| `age#3`       | ❌ No     | `#` not allowed           |
| `first-name`  | ❌ No     | `-` not allowed           |
| `my name`     | ❌ No     | Space not allowed         |
| `price$`      | ❌ No     | `$` not allowed           |


In [48]:
abc$xyz=6

SyntaxError: invalid syntax (3943422283.py, line 1)

interview question

# Let's try to find an area of a trainle using python. Now we all know this basic math that area of a triangle is ,


area = (1/2) * base * height

In [1]:
base = 10
height = 7

In [2]:
area = (1/2) * base * height
area

35.0

In [3]:
type(area)

float

# Two type of numbers: Integers and float 

In [4]:
weight = 100.5
type(weight)

float

In [7]:
family_members = 5
type(5)

int

# Basic airthmetic with numbers

In [8]:
x = 10
y = 3
x,y

(10, 3)

In [9]:
x + y

13

In [10]:
x - y

7

In [11]:
x/y

3.3333333333333335

In [12]:
x//y

3

In [13]:
x%y

1

In [14]:
x**y

1000

In [15]:
10+2*3

16

In [16]:
(10+2)*3

36

In [17]:
type(5)

int

In [18]:
type(5.2)

float

In [20]:
foo=2.3e4
foo

23000.0

In [21]:
1e4

10000.0

In [25]:
bar=2.3e-3
bar

0.0023

In [26]:
2.3*(10**-3)

0.0023

In [27]:
food = 120.34
rent = 500.4
utilities = 200.5


In [28]:
total = food + rent + utilities
total

821.24

In [29]:
round(total)

821

In [30]:
income = 10000
expenses = 9000
savings = income - expenses
savings

1000

In [31]:
food = "120.34"
rent = "500.67"

total = food + rent
total

'120.34500.67'

In [32]:
food = "120.34"
rent = "500.67"

# Convert strings to floats
total = float(food) + float(rent)
total

621.01

# Using Math Functions

In [33]:
import math
x = 16
math.sqrt(16)

4.0

In [34]:
math.floor(3.7)

3

In [36]:
math.ceil(3.7)

4

In [37]:
math.pi

3.141592653589793

In [38]:
dir(math)

['__doc__',
 '__loader__',
 '__name__',
 '__package__',
 '__spec__',
 'acos',
 'acosh',
 'asin',
 'asinh',
 'atan',
 'atan2',
 'atanh',
 'cbrt',
 'ceil',
 'comb',
 'copysign',
 'cos',
 'cosh',
 'degrees',
 'dist',
 'e',
 'erf',
 'erfc',
 'exp',
 'exp2',
 'expm1',
 'fabs',
 'factorial',
 'floor',
 'fma',
 'fmod',
 'frexp',
 'fsum',
 'gamma',
 'gcd',
 'hypot',
 'inf',
 'isclose',
 'isfinite',
 'isinf',
 'isnan',
 'isqrt',
 'lcm',
 'ldexp',
 'lgamma',
 'log',
 'log10',
 'log1p',
 'log2',
 'modf',
 'nan',
 'nextafter',
 'perm',
 'pi',
 'pow',
 'prod',
 'radians',
 'remainder',
 'sin',
 'sinh',
 'sqrt',
 'sumprod',
 'tan',
 'tanh',
 'tau',
 'trunc',
 'ulp']

# Binary Representation of integers

In [39]:
format(5,'b')

'101'

# DETAILS

# WHY?

5 is a decimal number.

'b' tells Python to convert it to binary.

Binary of 5 = 101

### CONVERT 5(decimal) = binary
Step 1: Divide the number by 2
Keep dicvinding by 2 and write the reminder each time.

Step	Divide by 2	Quotient	Remainder

1	5 ÷ 2	2	1

2	2 ÷ 2	1	0

3	1 ÷ 2	0	1

# Final Answer


Binary of 5 =101

In [41]:
# more examples
format(1,'b')

'1'

In [42]:
format(7,'b') 

'111'

In [43]:
format(10,'b')

'1010'

# math function
 * Constants

pi = 3.14159

e = 2.71828

tau = 2π

inf,nan

 
 * Common functions

sqrt(x) = square root

factorial(x)

pow(x,y)

log(x) , log10(x) , log2(x)

sin, cos , tan

degrees(),radians()

ceil(), floor(), trunc()

In [3]:
# 1 import math
import math

In [4]:
# 2. Trigonometry functions
print(math.sin(0))
print(math.cos(0))
print(math.tan(0))

print(math.asin(1))
print(math.acos(1))
print(math.atan(1))
print(math.atan2(1,1))

0.0
1.0
0.0
1.5707963267948966
0.0
0.7853981633974483
0.7853981633974483


In [5]:
# 3 . Hyperbolic functions
print(math.sinh(1))
print(math.cosh(1))
print(math.tanh(1))

print(math.asinh(1))
print(math.acosh(2))
print(math.atanh(0.5))

1.1752011936438014
1.5430806348152437
0.7615941559557649
0.881373587019543
1.3169578969248166
0.5493061443340549


In [6]:
# 4 Rounding Function
print(math.ceil(5.2))
print(math.floor(5.8))
print(math.trunc(5.9))

6
5
5


In [8]:

# 5 Logarithms
print(math.log(10))       # Natural logarithm (base e)
print(math.log10(10))     # Base 10 logarithm
print(math.log2(8))       # Base 2 logarithm
print(math.log1p(1))      # log(1 + x)


2.302585092994046
1.0
3.0
0.6931471805599453


In [9]:
# 6 Powers and Roots
print(math.sqrt(25))
print(math.pow(2, 3))
print(math.exp(1))
print(math.expm1(1))
print(math.exp2(3))

5.0
8.0
2.718281828459045
1.718281828459045
8.0


In [10]:
# 7 Factorial, GCD , Permutation , Combination
print(math.factorial(5))
print(math.gcd(12,8))
print(math.perm(5,2))
print(math.comb(5,2))

120
4
20
10


In [11]:
# 8 other useful Functions
print(math.fabs(-5))
print(math.fmod(7, 3))
print(math.modf(5.75))
print(math.copysign(5, -1))
print(math.dist([0,0],[3,4]))
print(math.hypot(3,4))
print(math.fsum([1,2,3]))
print(math.prod([2,3,4]))
print(math.remainder(10,3))

5.0
1.0
(0.75, 5.0)
-5.0
5.0
5.0
6.0
24
1.0


In [12]:
# 9 Constants
print(math.pi)
print(math.e)
print(math.tau)
print(math.inf)
print(math.nan)

3.141592653589793
2.718281828459045
6.283185307179586
inf
nan


In [13]:
# 10 Checking Numbers
print(math.isfinite(5))
print(math.isinf(math.inf))
print(math.isnan(math.nan))
print(math.isclose(1.00001, 1.0))

True
True
True
False


 ## To Be Continuee